### PR TITLE
Concat

### PR DESCRIPTION
**Pull request by KentonMurray**
This is a transformation that concats training data examples to create longer examples. This has been shown to be effective for Low-Resource Machine Translation by Nguyen et al., 2021

### CODE REVIEW

File: transformations/concat/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Concatenate Two Random Sentences
+This transformation concats two random sentences as explained in [Nguyen et al., 2021](https://arxiv.org/abs/2105.01691). This is the bilingual version. A monolingual version is available under concat monlingual.
+
```
Comment (Nickeilf) [inline transformations/concat/README.md:3]: Could you please add your information here, name/contact/institution, etc.?

File: transformations/concat_monolingual/transformation.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
+import re
+
+from checklist.perturb import Perturb
+
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+
+class ConcatMonolingual(SentenceOperation):
```
Comment (Nickeilf) [inline transformations/concat_monolingual/transformation.py:9]: I'm not sure about the goal of having a monolingual concat transformation. Is it intended for classification tasks?

I can understand that concating random source + translation might work as an augmentation method where model can learn from more combinations of data. However, for the monolingual concat, if only the source is concatenated with a random sentence, but the target (either a label in the classification task or a sentence in translation task) remains unchanged, how would it help the model? I would imagine there can be a meaning mismatch in this way.

**Review by Nickeilf — COMMENTED** [review]
Hi, @KentonMurray! This transformation looks good and interesting. However, it works on corpus-level rather than sentence-level (which means that the most appropriate interface is not found at the moment). Therefore, the current implementation uses two variables to cache the previous sample and add it to the current sample. I think the best way to do it is to first load a parallel corpus, and then randomly sample from the corpus to do the concatenation. That might introduce more diverse results than simply taking the previous sample.

We currently have a PR under review, which added the corpus-level interface to this repo. I'd suggest let's wait for that PR to be merged, and then you can change this transformation from `SentenceOperation` to the new corpus-level operation.

@kaustubhdhole What's your opinion?

**Comment by KentonMurray** [comment]
It's true that this operates on a corpus level - however previously published results (which this is based off of, Nguyen et al., 2021) have shown that concatenating two adjacent sentences works just as well as random. That paper actually surprised people because concatenating consecutive sentences had been claimed to use discourse information. However it didn't do any better than random. The benefit comes from just seeing variable length sentences, differing contexts, and positional information. All of these actually exist with adjacent sentences as well.

**Comment by Nickeilf** [comment]
> It's true that this operates on a corpus level - however previously published results (which this is based off of, Nguyen et al., 2021) have shown that concatenating two adjacent sentences works just as well as random. That paper actually surprised people because concatenating consecutive sentences had been claimed to use discourse information. However it didn't do any better than random. The benefit comes from just seeing variable length sentences, differing contexts, and positional information. All of these actually exist with adjacent sentences as well.

Thanks very interesting, and I have no more questions about this transformation. Could you please add your name/email/affiliation in the README?

File: transformations/concat/README.md (hunk 1 above)
Comment (KentonMurray) [reply transformations/concat/README.md:3]: Done. Thanks!

**Comment by KentonMurray** [comment]
@kaustubhdhole I only see one reviewer has been assigned. Does this need a second?

**Review by mille-s — APPROVED** [review]

**Comment by KentonMurray** [comment]
I've attached the error message from the build. This is something new and appears to deal with "transformation.ChangeDateFormat" which is not a transformation that I created. Basically, expected 2019 vs. predicted 2020, in the output. Is this test causing errors for anyone else? It seems to be independent of my PR


`=================================== FAILURES ===================================
_________________________ test_operation[light-light] __________________________

transformation_name = 'light', filter_name = 'light'

    def test_operation(transformation_name, filter_name):
        initialize_models()
>       execute_test_case_for_transformation(transformation_name)

test/test_main.py:154: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_main.py:125: in execute_test_case_for_transformation
    execute_sentence_operation_test_case(transformation, test)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

transformation = <transformations.change_date_format.transformation.ChangeDateFormat object at 0x7f0fb42fdb50>
test = {'class': 'ChangeDateFormat', 'inputs': {'sentence': 'The first known case of COVID-19 was identified in Wuhan, China ...ber 2019.'}, 'outputs': [{'sentence': 'The first known case of COVID-19 was identified in Wuhan, China in Dec 2019.'}]}

    def execute_sentence_operation_test_case(transformation, test):
        filter_args = test["inputs"]
        outputs = test["outputs"]
        perturbs = transformation.generate(**filter_args)
        for pred_output, output in zip_longest(perturbs, outputs):
>           assert pred_output == output["sentence"], get_assert_message(
                transformation, output["sentence"], pred_output
            )
E           AssertionError: Mis-match in expected and predicted output for ChangeDateFormat transformation: 
E              Expected Output: The first known case of COVID-19 was identified in Wuhan, China in Dec 2019. 
E              Predicted Output: The first known case of COVID-19 was identified in Wuhan, China in Dec 2020.
E           assert 'The first kn... in Dec 2020.' == 'The first kn... in Dec 2019.'
E             Skipping 63 identical leading characters in diff, use -v to show
E             -  in Dec 2019.
E             ?           ^^
E             +  in Dec 2020.
E             ?           ^^

test/test_main.py:33: AssertionError`

**Comment by kaustubhdhole** [comment]
Okay, I think @AbinayaM02 might have a better clue for that. I will let this be merged meanwhile since the build in main is also failing due to ChangeDate

**Review by kaustubhdhole — APPROVED** [review]

**Comment by KentonMurray** [comment]
Got it. Thanks.

**Pull request merged**